graphbolt/src/cuda/utils.h

```diff
@@ -51,6 +51,31 @@ int NumberOfBits(const T& range) {
   return bits;
 }
 
+/**
+ * @brief Given a sorted array and a value this function returns the index
+ * of the first element which compares greater than or equal to value.
+ *
+ * This function assumes 0-based index
+ * @param A: ascending sorted array
+ * @param n: size of the A
+ * @param x: value to search in A
+ * @return index, i, of the first element st. A[i]>=x. If x>A[n-1] returns n.
+ * if x<A[0] then it returns 0.
+ */
+template <typename indptr_t, typename indices_t>
```

Comment:
is there any utility in cuda/thrust/cccl library instead of implementing on your own?

Reply:
I checked quickly but wasn't able to find it.

Reply:
https://thrust.github.io/doc/group__binary__search_ga19da452d0b760bd927865ad2e0f45e41.html#ga19da452d0b760bd927865ad2e0f45e41 ?

Reply:
I was using this documentation website. https://nvidia.github.io/cccl/thrust/api.html

Reply:
I will try and see if they work. If it works, I will replace UpperBound as well.

Reply:
I think we are not supposed to use them. I got errors like this:
```
/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const unsigned char *, unsigned char> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<unsigned char, unsigned char> ::operator ()") is not allowed

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const signed char *, signed char> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<unsigned char, signed char> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const int *, int> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<unsigned char, int> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const long *, long> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<unsigned char, long> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const short *, short> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<unsigned char, short> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const unsigned char *, unsigned char> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<signed char, unsigned char> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const signed char *, signed char> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<signed char, signed char> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const int *, int> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<signed char, int> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const long *, long> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<signed char, long> ::operator ()") is not allowed

/home/mfbalin/dgl-1/graphbolt/src/cuda/sampling_utils.cu(83): warning #20011-D: calling a __host__ function("T1 thrust::THRUST_200300_890_NS::lower_bound<const short *, short> (T1, T1, const T2 &)") from a __host__ __device__ function("graphbolt::ops::EdgeTypeSearch<signed char, short> ::operator ()") is not allowed
```